cdap-data-fabric/src/main/java/co/cask/cdap/data2/dataset2/lib/partitioned/PartitionedFileSetDataset.java

```diff
@@ -186,29 +239,30 @@ public PartitionOutput getPartitionOutput(PartitionKey key) {
   }
 
   @Override
-  public Partition getPartition(PartitionKey key) {
-    final byte[] rowKey = generateRowKey(key, partitioning);
+  public PartitionDetail getPartition(PartitionKey key) {
+    byte[] rowKey = generateRowKey(key, partitioning);
     Row row = partitionsTable.get(rowKey);
     if (row == null) {
       return null;
     }
+
     byte[] pathBytes = row.get(RELATIVE_PATH);
     if (pathBytes == null) {
       return null;
     }
-    return new BasicPartition(Bytes.toString(pathBytes), key);
+    return new BasicPartitionDetail(this, Bytes.toString(pathBytes), key, metadataFromRow(row));
   }
 
   @Override
-  public Set<Partition> getPartitions(@Nullable PartitionFilter filter) {
-    final Set<Partition> partitions = Sets.newHashSet();
+  public Set<PartitionDetail> getPartitions(@Nullable PartitionFilter filter) {
+    final Set<PartitionDetail> partitionDetails = Sets.newHashSet();
     getPartitions(filter, new PartitionConsumer() {
       @Override
-      public void consume(PartitionKey key, String path) {
-        partitions.add(new BasicPartition(path, key));
+      public void consume(PartitionKey key, String path, @Nullable PartitionMetadata metadata) {
+        partitionDetails.add(new BasicPartitionDetail(PartitionedFileSetDataset.this, path, key, metadata));
```

Comment:
PartitionDetail assumes that Metadata is never null, right? But here metadata is @Nullable. I guess if it's null, you would want to use an empty meta data object?